encodings/alp/src/array.rs

```diff
@@ -109,7 +112,121 @@ impl ArrayVariants for ALPArray {
     }
 }
 
-impl PrimitiveArrayTrait for ALPArray {}
+struct ALPAccessor<F: ALPFloat> {
+    encoded: Arc<dyn Accessor<F::ALPInt>>,
+    patches: Option<Arc<dyn Accessor<F>>>,
+    validity: Validity,
+    exponents: Exponents,
+}
+impl<F: ALPFloat> ALPAccessor<F> {
+    fn new(
+        encoded: AccessorRef<F::ALPInt>,
+        patches: Option<AccessorRef<F>>,
+        exponents: Exponents,
+        validity: Validity,
+    ) -> Self {
+        Self {
+            encoded,
+            patches,
+            validity,
+            exponents,
+        }
+    }
+}
+
+impl<F: ALPFloat> Accessor<F> for ALPAccessor<F> {
+    fn array_len(&self) -> usize {
+        self.encoded.array_len()
+    }
+
+    fn is_valid(&self, index: usize) -> bool {
+        self.validity.is_valid(index)
+    }
+
+    fn value_unchecked(&self, index: usize) -> F {
+        match self.patches.as_ref() {
+            Some(patches) if patches.is_valid(index) => patches.value_unchecked(index),
+            _ => {
+                let encoded = self.encoded.value_unchecked(index);
+                F::decode_single(encoded, self.exponents)
+            }
+        }
+    }
+
+    fn array_validity(&self) -> Validity {
+        self.validity.clone()
+    }
+
+    fn decode_batch(&self, start_idx: usize) -> Vec<F> {
+        let mut values = self
+            .encoded
+            .decode_batch(start_idx)
+            .into_iter()
+            .map(|v| F::decode_single(v, self.exponents))
+            .collect::<Vec<F>>();
```

Comment:
we should not create a new vec here but instead preallocate a vec in the accessor and reuse it

Reply:
I verified this case doesn't allocate a new vec, but I agree that we should probably have a preallocated vec and try to return a `&` to it

Reply:
Right, I found on rust forums that vec.into_iter()...collect() will make effort to reuse memory